File: src/com/jacamars/dsp/rtb/shared/Customer.java
package com.jacamars.dsp.rtb.shared;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import com.hazelcast.nio.serialization.Portable;
import com.hazelcast.nio.serialization.PortableReader;
import com.hazelcast.nio.serialization.PortableWriter;

public class Customer implements Portable {
    public static final int CLASS_ID = 1;

    public String name;
    public int id;
    public Date lastOrder;
    public List<String> list;

    @Override
    public int getFactoryId() {
        return SamplePortableFactory.FACTORY_ID;
    }

    @Override
    public int getClassId() {
        return CLASS_ID;
    }

    @Override
    public void writePortable(PortableWriter writer) throws IOException {
        writer.writeInt("id", id);
        writer.writeUTF("name", name);
        writer.writeLong("lastOrder", lastOrder.getTime());
        var size = 0;
        if (list != null) {
        	size = list.size();
        }
       
        writer.writeInt("size", size);
        for (var i=0;i<size;i++) {
        	writer.writeUTF("list:"+i,list.get(i));
        }
    }

    @Override
    public void readPortable(PortableReader reader) throws IOException {
        id = reader.readInt("id");
        name = reader.readUTF("name");
        lastOrder = new Date(reader.readLong("lastOrder"));
        var size = reader.readInt("size");
        for (var i=0;i<size;i++) {
        	if (list == null)
        		list = new ArrayList();
        	list.add(reader.readUTF("list:"+i));
        }
    }
}

File: nodejs/customer.js
var Client = require('hazelcast-client').Client;
var Config = require('hazelcast-client').Config;
var Long = require('long');

function Customer(name, id, lastOrder, list) {
    this.name = name;
    this.id = id;
    this.lastOrder = lastOrder;
    this.list = list;
}

Customer.prototype.readPortable = function (reader) {
    this.name = reader.readUTF('name');
    this.id = reader.readInt('id');
    this.lastOrder = reader.readLong('lastOrder').toNumber();
    let size = reader.readInt("size");
    this.list = [];
    for (let i=0; i<size;i++) {
	let s = reader.readUTF("list:"+i);
	this.list.push(s);
    }
};

Customer.prototype.writePortable = function (writer) {
    writer.writeUTF('name', this.name);
    writer.writeInt('id', this.id);
    writer.writeLong('lastOrder', Long.fromNumber(this.lastOrder));
};

Customer.prototype.getFactoryId = function () {
    return 1;
};

Customer.prototype.getClassId = function () {
    return 1;
};

function PortableFactory() {
    // Constructor function
}

PortableFactory.prototype.create = function (classId) {
    if (classId === 1) {
        return new Customer();
    }
    return null;
};

let cfg = new Config.ClientConfig();
cfg.serializationConfig.portableFactories[1] = new PortableFactory();
// Start the Hazelcast Client and connect to an already running Hazelcast Cluster on 127.0.0.1

getCustomer(cfg);

async function getCustomer(config) {
	let client = await Client.newHazelcastClient(config);
        let customers = await client.getMap('CUSTOMERS');
	let customer = await customers.get("Ben");
	console.log("Customer: " + JSON.stringify(customer));
}
